### PR TITLE
Await subscribes before unsubscribing

### PR DESCRIPTION
**Pull request by fb55**
... and return unsubscribe promise.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -52,13 +52,16 @@ module.exports = class SubscriptionDedupe {
 
   unsubscribe(topic/*: string */) {
     let existing = this.subscriptions[topic];
+    let promise
 
     if (existing) {
       existing.refCount--;
       if (existing.refCount === 0) {
-        this.options.onUnsubscribe(topic);
+        promise = existing.promise.then(() => this.options.onUnsubscribe(topic));
```
Comment (fb55) [inline index.js]: This can still have a race-condition if we rapidly reopen the subscription, will update.

Comment (STRML) [reply index.js]: Thanks. Could you add a test as well?

File: index.js [hunk 2]
```diff
@@ -35,30 +35,56 @@ module.exports = class SubscriptionDedupe {
...
         promise: this.options.onSubscribe(topic),
-        refCount: 0
+        refCount: 0,
+        closing: null,
       };
     }
 
+    if (existing.closing) {
+      existing.closing.isReopened = true;
+      existing.closing = null;
```
Comment (STRML) [inline index.js]: Won't this overwrite L51?

File: index.js [hunk 3]
```diff
@@ -35,30 +35,56 @@ module.exports = class SubscriptionDedupe {
...
       // Nothing found, create the subscription object
       existing = this.subscriptions[topic] = {
         promise: this.options.onSubscribe(topic),
-        refCount: 0
+        refCount: 0,
+        closing: null,
       };
     }
 
+    if (existing.closing) {
```
Comment (STRML) [inline index.js]: Clearer as an `else if`

File: index.js (hunk 2 above)
Comment (fb55) [reply index.js]: This is necessary to allow multiple pending reopens. The object is still part of the scope of the function deleting the subscription object in L82.

**Comment by STRML** [comment]
Thanks for all the help @fb55!

**Comment by STRML** [comment]
Published v1.2.0.

**Pull request merged**